### ISSUE TITLE
Use `dataclasses` for all objects returned by `HfApi`

### ISSUE THREAD

**Issue by mariosasko** [huggingface/huggingface_hub#1911]
Currently, we use `TypedDict` to represent some of the objects' attributes returned by the `HfApi` methods. To be consistent, we should only use `dataclasses`, as suggested in https://github.com/huggingface/huggingface_hub/pull/1809#discussion_r1387635398:

> I think we should open a separate PR to switch all of them (`BlobLfsInfo`, `LastCommitInfo`, `BlobSecurityInfo`, `TransformersInfo`, `SafeTensorsInfo`) to dataclasses, in a backward compatible way (and with a deprecation warning when dict-only method is used). This way we'll finally have a single type to represent data returned by the server (now that you've removed ReprMixin).

This work can be split in several PRs to make it easier to implement and review.

**Comment by NouamaneELGueddarii** [comment]
Hello mario, i would like to do this task, is it possible ?

**Comment by Wauplin** [comment]
Hi @NouamaneELGueddarii! Thanks for proposing yourself! It is possible to take this task yes. Just to be sure we are aligned, the goal is to update a bunch of objects in [`hf_api.py`](https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/hf_api.py) from `TypedDict` to proper dataclasses. At first, typed dict were used only for static type checking. However now we want to convert everything to dataclasses that have more features. So for each object, you would need to:
1. inherit from `dict` like [in this one](https://github.com/gradio-app/gradio/blob/8dfabee00495ccbbd6743da07fa06c75cac3fb5f/gradio/oauth.py#L143)
2. decorate with `@dataclass` like [in this one](https://github.com/gradio-app/gradio/blob/8dfabee00495ccbbd6743da07fa06c75cac3fb5f/gradio/oauth.py#L142)
3. accept a dict as input in `__init__.py` (see [here](https://github.com/gradio-app/gradio/blob/8dfabee00495ccbbd6743da07fa06c75cac3fb5f/gradio/oauth.py#L179))
4. update `self.__dict__` (see [here](https://github.com/huggingface/huggingface_hub/blob/66f8ea6264adecfa904b4c93944e9a230dc207c9/src/huggingface_hub/hf_api.py#L694)) for backward **and** future compatibility (if server response gets extended)
5. update how objects are instantiated

What I would suggest you is to start working on just 1 or 2 classes and open a PR for them. This way we can iterate on it and answer your questions (if any). What do you think?

Thanks in advance!

**Comment by Ahmedniz1** [comment]
@Wauplin is it fine if I start working on it? I want to start making open source contribution to github. 
Thanks

**Comment by Wauplin** [comment]
Hi @Ahmedniz1, yes it is! Thanks for proposing you. Please refer to https://github.com/huggingface/huggingface_hub/issues/1911#issuecomment-1881212204 on how to deal with it.
Since @NouamaneELGueddarii mentioned earlier they would like to work on it as well, could you share on which dataclass you would like to focus first before starting implementing something?  A simple message on this thread is enough (just want to avoid duplicate work).

> I want to start making open source contribution to github.

:rocket: :rocket: :rocket:

**Comment by Wauplin** [comment]
Sounds like a plan!

**Comment by NouamaneELGueddarii** [comment]
@Wauplin i will be working on the following class "TransformersInfo".  I will make a PR and wait for approval then proceed to work on the others.

**Comment by Wauplin** [comment]
Great! Thanks for confirming @NouamaneELGueddarii!

**Comment by Wauplin** [comment]
@Ahmedniz1, saw it and just reviewed it! Thanks!

**Comment by Wauplin** [comment]
@Ahmedniz1 @NouamaneELGueddarii I have updated the instructions on how to proceed with the issue. I'm sorry about this late change but I realized the previous recommendations were working correctly. I've commented on @Ahmedniz1's PR here (https://github.com/huggingface/huggingface_hub/pull/1974#pullrequestreview-1827419034) and updated the comment above (https://github.com/huggingface/huggingface_hub/issues/1911#issuecomment-1881212204). Hopefully, this will make it easier to integrate new dataclasses while keeping backward compatibility.
Sorry about the change and thanks in advance for adapting!

**Comment by Ahmedniz1** [comment]
@Wauplin I'm trying to update in the previous PR but it appears to be closed. I can push the updates for other classes(apart from transformerinfo) once the original PR is open again.
Thanks

**Comment by Wauplin** [comment]
> I'm trying to update in the previous PR but it appears to be closed.

Yes this is expected! You have to checkout to `main`, pull the latest changes, create a new branch from `main` and start a separate PR from there. This way we have small PRs that can easily be reviewed. The work you've already done is now on the `main` branch.
